ui/radar/radar.js

```diff
@@ -0,0 +1,281 @@
+'use strict';
+
+let Options = {
+  Language: 'en',
+  SplitSameMonster: false,
+  UpdateMonsterPos: false,
+  DetectionRange: 0,
+  OnlyMobs: true,
+  TTS: false,
+  PopSoundAlert: true,
+  PopSound: '../../resources/sounds/PowerAuras/sonar.ogg',
+  PopVolume: 0.5,
+};
+
+let gRadar;
+
+let instanceChangedRegex = {
+  'en': /00:0039:You are now in the instanced area/,
+  'cn': /00:0039:当前所在副本区为/,
+  'ja': /00:0039:インスタンスエリア/,
+  'de': /00:0039:Du bist nun in dem instanziierten Areal/,
+  'fr': /00:0039:Vous avez été transporté/,
+};
+
+class Point2D {
+  constructor(x, y) {
+    this.x = x;
+    this.y = y;
+  }
+}
+
+function length(a) {
+  return Math.sqrt((a.x) * (a.x) + (a.y) * (a.y));
+}
+
+function dot(a, b) {
+  return a.x * b.x + a.y * b.y;
+}
+
+function theta(a, b) {
+  return dot(a, b)/(length(a) * length(b));
+}
+
+function calc_damage(x) {
+  if (x.length < 8)
+    x = new Array(9 - x.length).join('0') + x;
+  x = '0x' + x;
+  let y = Number(x);
+  if (y & 0x00004000) {
+    // lot of damage
+  } else {
+    y = y & 0x77770000;
+    y >>= 16;
+  }
+  return y;
+}
+
+class Radar {
+  constructor(element) {
+    this.targetMonsters = [];
+    this.playerPos = {};
+    this.table = element;
+    this.options = Options;
+    this.monsters = Object.assign({}, gMonster, Options.CustomMonsters);
+  }
+
+  OnLogEvent(e) {
+    let lang = this.options.Language;
+    for (let i = 0; i < e.detail.logs.length; i++) {
+      // add new combatant
+      let r = e.detail.logs[i].match(/03:(........):Added new combatant (.*)\. (?:.*) Max HP: (\d+) (?:.*) Pos: \((-?\d+.\d+),(-?\d+.\d+),(-?\d+.\d+)\)/);
+      if (r) {
+        for (let name in this.monsters) {
+          let r_id = r[1];
+          let r_name = r[2];
+          let r_hp = r[3];
+          let r_pos_x = r[4];
+          let r_pos_y = r[5];
+          let r_pos_z = r[6];
+          let monster = this.monsters[name];
+          let matchOrNot = (r_name.match(monster['name'][lang] || monster['name']['en']) != null);
+          matchOrNot &= (r[0].match(monster['regex'] || '') != null);
+          matchOrNot &= (Number(r_hp) >= (monster['hp'] || 0));
```

Comment:
Chinese players may want to use the minimum hp to filter out chocobos, which may have the same name as hunt mobs and also starts with 4 in the ObjectID of log.
I’m not sure whether this will happen in other servers so I didn’t remove the hp filter, but that’s compatible with current hunt resources (without any hp info) I think.
monster[‘regex’] is another leftover filter setting for user to custom, maybe I should mention it in the comment lol.